repo2docker/engine.py

```diff
@@ -0,0 +1,308 @@
+        """
+
+
+class Image:
+    """
+    Information about a container image
+    """
+
+    def __init__(self, *, tags, config=None):
+        self._tags = tags or []
+        self._config = config
+
+    @property
+    def tags(self):
+        """
+        A list of tags associated with an image.
+
+        If locally built images have a localhost prefix this prefix should be removed or the image may not be recognised.
+        If there are no tags [] will be returned.
+        """
+        return self._tags
+
+    @property
+    def config(self):
+        """
+        A dictionary of image configuration information
+
+        If this is `None` the information has not been loaded.
+        If not `None` this must include the following fields:
+        - WorkingDir: The default working directory
+        """
+        return self._config
+
+    def __repr__(self):
+        return "Image(tags={},config={})".format(self.tags, self.config)
+
+
+class ContainerEngine(LoggingConfigurable):
+    """
+    Abstract container engine.
+
+    Inherits from LoggingConfigurable, which means it has a log property.
+    Initialised with a reference to the parent so can also be configured using traitlets.
+    """
+
+    string_output = True
+    """
+    Whether progress events should be strings or an object.
+
+    Originally Docker was the only container engine supported by repo2docker.
+    Some operations including build() and push() would return generators of events in a Docker specific format.
+    This format of events is not easily constructable with other engines so the default is to return strings and raise an exception if an error occurs.
+    If an engine returns docker style events set this variable to False.
+    """
+
+    def __init__(self, *, parent):
+        """
+        Initialise the container engine
+
+        Parameters
+        ----------
+        parent: Application
+            Reference to the parent application so that its configuration file can be used in this plugin.
+        """
+        super().__init__(parent=parent)
+
+    # Based on https://docker-py.readthedocs.io/en/4.2.0/api.html#module-docker.api.build
+
+    def build(
+        self,
+        *,
+        buildargs={},
+        cache_from=[],
+        container_limits={},
+        tag="",
+        custom_context=False,
+        dockerfile="",
+        fileobj=None,
+        path="",
+        **kwargs
+    ):
+        """
+        Build a container
+
+        Parameters
+        ----------
+        buildargs : dict
+            Dictionary of build arguments
+        cache_from : list[str]
+            List of images to chech for caching
+        container_limits : dict
+            Dictionary of resources limits. These keys are supported:
+              - `cpusetcpus`
+              - `cpushares`
+              - `memory`
+              - `memswap`
+        tag : str
+            Tag to add to the image
+
+        custom_context : bool
```

Comment:
`custom_context` and `fileobj` and very biased towards the Docker API. However it's fairly easy in Python to untar this tarfile-like-object into a temporary directory so it can be made to work.